### ISSUE TITLE
Semantic Syntax is incorrectly highlighting arguments as commands

### ISSUE THREAD

**Issue by Jaykul** [PowerShell/vscode-powershell#2860]
Any bare (unquoted) string argument that has a `-` in it is being highlighted as a command:

![image](https://user-images.githubusercontent.com/192942/89322654-89748a80-d652-11ea-9763-c927099b5c89.png)

**Comment by justinytchen** [comment]
The problem is that there is no way to distinguish between a function name in a declaration and a parameter from just the token information alone. I think that we'll need to add other checks (i.e. position or looking into the AST)

![image](https://user-images.githubusercontent.com/6785106/89453594-5abee880-d714-11ea-9aae-e7dd1307d88b.png)

![image](https://user-images.githubusercontent.com/6785106/89453539-48dd4580-d714-11ea-8e6e-3ac1f246ce42.png)

**Comment by rjmholt** [comment]
I think the important part is that the `Test` token and the `That-Thing` token have different flags:

```
> $ast = [System.Management.Automation.Language.Parser]::ParseInput('get-banana this-thing', [ref]$tok
s, [ref]$errs)
C:\Users\Robert Holt\Documents\Dev\Microsoft\PowerShellEditorServices [async-ps-consumer +0 ~4 -0 !]
> $toks[0].TokenFlags
CommandName
C:\Users\Robert Holt\Documents\Dev\Microsoft\PowerShellEditorServices [async-ps-consumer +0 ~4 -0 !]
> $toks[1].TokenFlags
None
```

That's how PSReadLine is able to differentiate them I imagine

**Comment by SeeminglyScience** [comment]
@rjmholt I think what they're getting at is that the current regex based highlighter treats the name in a function definition like a command (which doesn't have the flag).

Personally I'd rather have the name in a function definition be parsed as a bare word than the current situation.  But a fix for both might be a flag of "was last token kind `Function`".  Then also use `TokenFlags.CommandName`.

**Comment by justinytchen** [comment]
There's no way to differentiate between the Get-A*A token (should be labelled as a command/function) and the That-Thing token (which should be labelled as something else) with only the Token info. So without using additional data, we either have the situation we have now, or Get-A*A would not be highlighted

**Comment by SeeminglyScience** [comment]
@justinytchen You also change this up a bit:

https://github.com/PowerShell/PowerShellEditorServices/blob/c40734354631043cf3c390d36c9b1ef6a80b3949/src/PowerShellEditorServices/Services/TextDocument/Handlers/PsesSemanticTokensHandler.cs#L51-L54

to be:

```csharp
Token[] tokens = file.ScriptTokens;
for (int i = 0; i < tokens.Length; i++)
{
    Token token = tokens[i];
    PushToken(token, builder);
    if (token.Kind == TokenKind.Function)
    {
        // + bounds checks
        Token next = tokens[++i];
        PushTokenAsFunctionNameIfBareWord(next, builder);
    }
}
```

**Comment by rjmholt** [comment]
> I think what they're getting at is that the current regex based highlighter treats the name in a function definition like a command (which doesn't have the flag).

Yes, I noticed that. But in @Jaykul's original issue he seems to be referring to the `CommandAst` situation rather than the `FunctionDefinitionAst` situation.

The `FunctionDefinitionAst` scenario is easy to address: we look back a token and see if it's the `function` keyword. Given that the grammar defines the full nature of the context, rather than just the tokens, we can't do a simple token search for all tokens, but I believe we can in this case.

A more sophisticated approach is to track our position in both the AST and the token array simulataneously and for each token look into where we are in the AST to determine the token semantics.

**Comment by rjmholt** [comment]
Ok so the problem is actually that generic tokens with no flags are [currently highlighted as functions](https://github.com/PowerShell/PowerShellEditorServices/blob/c40734354631043cf3c390d36c9b1ef6a80b3949/src/PowerShellEditorServices/Services/TextDocument/Handlers/PsesSemanticTokensHandler.cs#L155-L156).

Bareword arguments without `-`s are tokenised as identifiers rather than generic tokens, so they fall through to the grammar.

We can recognise a bareword identifier most correctly as an identifier or a generic token that is subordinate to a command token... Harder to do though.

We need to pick a new value based on [this list](https://github.com/OmniSharp/csharp-language-server-protocol/blob/f93f973ec96d895c0c6056f5f5942c4646071ab9/src/Protocol/Models/Proposals/SemanticTokenTypes.cs#L35-L57). I would guess `Member`, `Property` or `EnumMember` would work.

**Comment by TylerLeonhardt** [comment]
Property is probably the least threatening

**Comment by rjmholt** [comment]
> because Member in the Monokai theme

Ugh, yeah, this is what I mean by themes requiring hacks

**Comment by Jaykul** [comment]
A "generic token" that's a parameter ... is actually a `string`, not a `member`.

**Comment by SeeminglyScience** [comment]
@Jaykul true but nothing currently highlights a bare word string literal command arg like every other string literal.  `member` is probably the closest to what existing tools highlight it as.

**Comment by Jaykul** [comment]
Are you really arguing that since everyone else is wrong, this should be wrong too? It's a string. The fact that PowerShell _allows_ you to leave off the quotes doesn't change the fact that it's a string. It should be highlighted as a string if it's `Kind = "Identifier"` or `Kind = "Generic"` with `TokenFlags = "None"` (and follows a command?).

It's **most definitely _not_** a member.  Highlighting it as member is _worse than nothing_ (although not worse than highlighting it as a command) in my opinion. 

If you really decide you _need_ to use member, please instead send a PR to the `SemanticTokenTypes` enum to add an a "unspecified" or "otherliteral" or something.

**Comment by SeeminglyScience** [comment]
> Are you really arguing that since everyone else is wrong, this should be wrong too?

Nah, I'm saying people will be annoyed if it's highlighted like every other string literal, because that's what they're used to.

> It's a string. The fact that PowerShell allows you to leave off the quotes doesn't change the fact that it's a string.

Yeah, no one here thinks it isn't a string.  Doesn't change the fact that folks expect it to be highlighted differently 🤷

**Comment by rjmholt** [comment]
Yeah it's not about how PowerShell uses it, it's about how it's highlighted. Categorising it as a string for highlighting purposes will make `'input'` and `input` the same colour, whereas in PSReadLine for example, there are not at all (`here` vs `'there'`):

![image](https://user-images.githubusercontent.com/7009879/89693193-a7482680-d8c2-11ea-83e3-976b7630d39e.png)

> Are you really arguing that since everyone else is wrong, this should be wrong too? It's a string.

Please keep your tone civil. We can be more constructive without rhetoric like this.

**Comment by Jaykul** [comment]
I don't think there's that much backward compatibility to worry about. It's already **drastically** changing most of the highlighting... and not just breaking changes like using statements with paths, escaped characters, `#requires`, etc.

**Comment by SeeminglyScience** [comment]
@Jaykul that's all fair and I don't necessarily disagree.  Tbh, I think it would bother people.  I don't really have any specific logic to back that up with, but I think it would get a lot of complaints).

Personally I'd also rather stick closer to what PSRL does, even if it isn't correct.  PSRL highlighting barewords differently changed how they feel vs quoted strings imo, for better or worse.

**Comment by rjmholt** [comment]
> Both your here and 'there' are strings. Why would you want to make them seem different?

Yeah, I definitely understand your reasoning here.

We're in a strange set because bareword strings are a rare occurrence among languages, so there's not good precedent or mechanisms built in (in an environment mainly designed to cater to C# and TypeScript) for them.

> I don't think there's that much backward compatibility to worry about

Agreed. Certainly as a visual feature, I don't personally see any concerns about "breaking changes" so to speak.

I think of highlighting as a visual representation of tokeniser mode. Even though bareword strings and quoted strings represent the same type of value, when I'm typing them or looking at my program, the thing I care about is what delimits them. A quote is easy to miss, but the specific colour of a quoted string tells me that I'm in a special mode where the rules are different. That's not true of bareword strings, because the tokeniser will find a lot more reasons to end the string token.

Certainly I think that if we made a change to classify quoted strings and bareword strings the same way, the majority of our users would have complaints due to struggling to differentiate the different types of strings in their scripts.

In an ideal world, we would have a `BareWordString` token type that a VSCode theme could pick up on. But in the real world we need a fallback that will cater to common themes out there that are mainly built for programming languages with tame syntaxes. I think that's the essence of my proposal in https://github.com/PowerShell/vscode-powershell/issues/2852#issuecomment-669259784.

(The unfortunate side-effect of the success of the C-syntax-family languages is that much tooling forgets just how different some languages are)

**Comment by Jaykul** [comment]
While you're convincing me this whole product is not for me, I disagree fundamentally with the base assumption that your users want what they already have.

Here's PSReadLine, Semantic, and EditorSyntax:

![image](https://user-images.githubusercontent.com/192942/89814137-b1f7fb00-db10-11ea-8d64-22d54493328d.png)

PSReadLine is better than the others, and **(when it's not coloring strings as commands)** sematic is marginally better than editor syntax, but the highlighting here is incorrect and _does not help authors understand their code._

I have no problem with the idea of having a language-specific token (or, in this case, a shell-language-specific token) and a fallback token type for themes that don't understand the type of language that you ware, but if you want to do that, you also need to provide a first-class theme that you can point to as an example (and in _this_ day and age, you need a dark version of it too). However, that's not what's happening here, and in this case, the ISE theme has never provided better information at all (if anything, it's worse), because **keywords** are virtually indistinquishable from bareword strings.

**Comment by TylerLeonhardt** [comment]
So I think we need to take this in two stages.

1. The first stage is **parity**: being as close to PSReadLine as possible. This will allow both the console and editing experience to be on the same playing field. (this was actually the original goal of doing this work from the beginning)

2. The second stage is **improving**: Once both experiences are roughly aligned, then we can go back and make the changes in both to support the language better.

**Comment by jessiewestlake** [comment]
To make matters worse... You also have the same problem when the parameter argument is not a string, but a negative number (e.g. `-1` or `-3.14`).

````PowerShell
Invoke-Calculation -Operator Subtract -Minuend -3 -Subtrahend 2
````
